### PR TITLE
Enable Strict Type Checking for Complex Data Types in check_contracts

### PR DESCRIPTION
**Pull request by CulmoneY**
## Proposed Changes

_(Describe your changes here. Also describe the motivation for your changes: what problem do they solve, or how do they improve the application or codebase? If this pull request fixes an open issue, [use a keyword to link this pull request to the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).)_
This PR enhances the @check_contracts decorator by extending strict numeric type checking across the entire numeric hierarchy, including support for complex numbers. Additionally, it introduces strict type validation for nested and union types, such as list[int], dict[float, int], and Union[int, float]. In the current behaviour, no error will be rasied, as in the example below.

```
@check_contracts
def foo(a: list[float]):
    print(a)

foo([1 , 2 , 3])
```
This demonstrates that, even though foo expects a list of floats, passing a list of integers does not raise an error when `STRICT_NUMERIC_TYPES` is enabled because of the relaxed type checking in nested contracts.

...

<details>
<summary>Screenshots of your changes (if applicable)</summary>

</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |          |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |    X      |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [x] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

### CODE REVIEW

File: python_ta/contracts/__init__.py [hunk 1]
```diff
@@ -321,17 +331,71 @@ def _check_function_contracts(wrapped, instance, args, kwargs):
 
 
 def check_type_strict(argname: str, value: Any, expected_type: type) -> None:
-    """Ensure that ``value`` matches ``expected_type``.
+    """
+    Ensure that `value` matches ``expected_type`` with strict type checking.
```
Comment (david-yz-liu) [inline python_ta/contracts/__init__.py:335]: Descriptions should always start on the same line as the triple-quotes (you should not change this). Make sure the functions you've added start with a short description after the triple-quotes.

File: python_ta/contracts/__init__.py [hunk 2]
```diff
@@ -321,17 +331,71 @@ def _check_function_contracts(wrapped, instance, args, kwargs):
...
 
-    Differentiates between:
-        - float vs. int
-        - bool vs. int
+    This function enforces strict type distinctions within the numeric hierarchy (bool, int, float,
+    complex), ensuring that the type of value is exactly the same as expected_type, not merely a subtype.
```
Comment (david-yz-liu) [inline python_ta/contracts/__init__.py:338]: Remove the phrase "not merely a subtype". Technically, `int` isn't a subclass of `float` (though `bool` _is_ a subclass of `int`), it's just that PEP 484 says type checkers can treat them as such.

File: python_ta/contracts/__init__.py [hunk 3]
```diff
@@ -321,17 +331,71 @@ def _check_function_contracts(wrapped, instance, args, kwargs):
...
+                _check_inner_type(argname, key, inner_types[0])
+                _check_inner_type(argname, item, inner_types[1])
+        else:
+            raise TypeError(f"type of {argname} must be {expected_type}; got {value} instead")
+    elif outer_type is tuple:
+        if isinstance(value, tuple) and len(inner_types) == 2 and inner_types[1] is Ellipsis:
+            for item in value:
+                _check_inner_type(argname, item, inner_types[0])
+        elif isinstance(value, tuple) and len(value) == len(inner_types):
+            for i, item in enumerate(value):
```
Comment (david-yz-liu) [inline python_ta/contracts/__init__.py:394]: use `zip`

File: tests/test_contracts/test_contracts.py [hunk 4]
```diff
@@ -853,3 +853,245 @@ def return_bool() -> bool:
...
+
+
+def test_parameter_tuple_fixed_length_types() -> None:
+    """Testing when the function expects a fixed-length tuple with specific types (tuple[int, str])."""
+
+    @check_contracts
+    def accept_fixed_tuple(value: tuple[int, str]) -> str:
+        return f"{value[0]}: {value[1]}"
+
+    accept_fixed_tuple((5, "test"))
```
Comment (david-yz-liu) [inline tests/test_contracts/test_contracts.py:1046]: separate these into separate test cases (like you did all other test cases; tuples aren't special)

**Comment by CulmoneY** [comment]
Hey @david-yz-liu. I added the test `test_custom_generic_class_with_parameter` to extend coverage to the else branch in _check_inner_type. This will make sure there are no uncovered lines, even though the else case is quite trival.

**Review by david-yz-liu — APPROVED** [review]

**Pull request merged**